### PR TITLE
Add Iceberg provider

### PR DESCRIPTION
**Pull request by romsharon98**
Following https://github.com/apache/airflow/pull/32786#issuecomment-1768284013 since Tabular provider doesn't have any specific tabular logic it has been suggested to move Tabular logic to new Apache Iceberg provider

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by eladkal** [comment]
@Fokko Do we want to keep Tabular provider (if there might be future tabular specific behavior?) or should we deprecate the whole tabular provider in favor of apache-iceberg provider? (difference between deprecate specific classes vs the whole provider)

**Comment by Fokko** [comment]
>  @Fokko Do we want to keep Tabular provider (if there might be future tabular specific behavior?) or should we deprecate the whole tabular provider in favor of apache-iceberg provider? (difference between deprecate specific classes vs the whole provider)

No, let's deprecate it in favor of the Iceberg provider. Everything in here is just Iceberg, and nothing Tabular specific, so the current naming is confusion

File: airflow/providers/apache/iceberg/hooks/iceberg.py
Comment (Fokko) [inline airflow/providers/apache/iceberg/hooks/iceberg.py]: Probably we want to get rid of this default :)

File: docs/apache-airflow-providers-apache-iceberg/connections.rst
Comment (Fokko) [inline docs/apache-airflow-providers-apache-iceberg/connections.rst]:
```suggestion
The Iceberg connection type enables connecting to an Iceberg REST catalog to request a short-lived token to access the Apache Iceberg tables. This token can be injected as an environment variable, to be used with Trino, Spark, Flink, or your favorite query engine that supports Apache Iceberg.
```

File: docs/apache-airflow-providers-apache-iceberg/connections.rst
Comment (Fokko) [inline docs/apache-airflow-providers-apache-iceberg/connections.rst]:
```suggestion
    The OAuth2 Client ID
```

File: docs/apache-airflow-providers-apache-iceberg/connections.rst
Comment (Fokko) [inline docs/apache-airflow-providers-apache-iceberg/connections.rst]:
```suggestion
    The OAuth2 Client Secret
```

**Comment by Fokko** [comment]
@romsharon98 Sorry for being late to the party here, but this looks great! Thanks for picking this up 🎉

File: airflow/providers/apache/iceberg/hooks/iceberg.py
Comment (romsharon98) [reply airflow/providers/apache/iceberg/hooks/iceberg.py]: Does iceberg have a default URL?

File: docs/apache-airflow-providers-tabular/redirects.txt
Comment (romsharon98) [inline docs/apache-airflow-providers-tabular/redirects.txt]: This redirect not working as expected.

File: airflow/providers/apache/iceberg/hooks/iceberg.py
Comment (Fokko) [reply airflow/providers/apache/iceberg/hooks/iceberg.py]: No, it needs to point to your catalog, so that differs per organization

**Review by Fokko — APPROVED** [review]
Some small comments, but this looks great @romsharon98

**Review by eladkal — CHANGES_REQUESTED** [review]
Marking as request changes just to wait for my review and after release of current provider wave
I'll get to it in 1-2 days

**Comment by Fokko** [comment]
@eladkal That makes sense. It was not my intent to merge this without your approval, but just to indicate that there are no further comments on the Operator and Hook itself.

**Comment by eladkal** [comment]
> @eladkal That makes sense. It was not my intent to merge this without your approval

I know :)
This is just to avoid merge by another committer. We are 99% ready here. it's just for the procedure of release due to https://github.com/apache/airflow/pull/39240
for simplicity I will handle it as ad-hoc release

File: airflow/providers/tabular/CHANGELOG.rst
Comment (Taragolis) [inline airflow/providers/tabular/CHANGELOG.rst]:
```suggestion
.. warning:: This is the last release for tabular provider. The provider is discontinued.
```

File: airflow/providers/tabular/CHANGELOG.rst
Comment (Taragolis) [inline airflow/providers/tabular/CHANGELOG.rst]:
```suggestion
   Please switch to `apache-airflow-providers-apache-iceberg <https://airflow.apache.org/docs/apache-airflow-providers-apache-iceberg/stable/index.html>`__ provider.
```

File: airflow/providers/tabular/__init__.py
Comment (Taragolis) [inline airflow/providers/tabular/__init__.py:1]: Need to revert changes in this file, it is generated during the release

File: docs/apache-airflow-providers-tabular/changelog.rst
Comment (eladkal) [inline docs/apache-airflow-providers-tabular/changelog.rst:1]: undo change to this file

File: docs/apache-airflow-providers-tabular/redirects.txt
Comment (eladkal) [reply docs/apache-airflow-providers-tabular/redirects.txt]: I think we don't need redirect here.
It's a different provider

File: airflow/providers/tabular/CHANGELOG.rst
Comment (eladkal) [inline airflow/providers/tabular/CHANGELOG.rst]:
```suggestion
```

No need for the version. Release manager auto pick up the comment at the change log and place it to the chosen release version

File: airflow/providers/tabular/provider.yaml
Comment (eladkal) [inline airflow/providers/tabular/provider.yaml]:
```suggestion
```

File: airflow/providers/tabular/hooks/tabular.py
Comment (eladkal) [inline airflow/providers/tabular/hooks/tabular.py:25]: In case where user updated tabular provider but did not install the iceberg provider.
In such case this import will fail. We need to wrap this with try/catch and raise exception (possibly `AirflowOptionalProviderFeatureException` ?)

**Review by eladkal — CHANGES_REQUESTED** [review]

File: docs/apache-airflow-providers-tabular/changelog.rst
Comment (romsharon98) [reply docs/apache-airflow-providers-tabular/changelog.rst:1]: done

File: airflow/providers/tabular/hooks/tabular.py
Comment (romsharon98) [reply airflow/providers/tabular/hooks/tabular.py:25]: Can it happen? because after those changes tabular depends on iceberg.

Comment (eladkal) [reply airflow/providers/tabular/hooks/tabular.py:25]: ah right

File: airflow/providers/tabular/.latest-doc-only-change.txt
Comment (eladkal) [inline airflow/providers/tabular/.latest-doc-only-change.txt]: I think this shouldn't be removed?

**Review by eladkal — APPROVED** [review]
LGTM
will merge when CI is green

File: airflow/utils/db.py
Comment (eladkal) [inline airflow/utils/db.py]: This is the reason for the test failures.
This file is sorted by conn_id changing tabular to icebreg break the sorting.
Note also that the host needs to be changed.

I think it's best to leave tabular and just add iceberg along side.
We can remove tabular conn_id if/when we remove the tabular provider

Comment (romsharon98) [reply airflow/utils/db.py]: 👑

**Review by potiuk — APPROVED** [review]

**Comment by eladkal** [comment]
🎉🎉🎉

**Comment by potiuk** [comment]
Wooohooo!

**Comment by Fokko** [comment]
Nice! Thanks @romsharon98 for working on this, and thanks @eladkal and @potiuk for the review 🎉

**Pull request merged**
